File: golang/src/backend-integration-impact/hello-server-token-binding-protected.go
// @link https://approov.io/docs/v2.1/approov-usage-documentation/#backend-integration-impact
// @link https://github.com/approov/documentation_backend-code-snippets/blob/master/golang/src/backend-integration-impact/hello-server-token-binding-protected.go

package main

import (
    "os"
    "fmt"
    "encoding/json"
    "encoding/base64"
    "log"
    "net/http"
    jwt "github.com/dgrijalva/jwt-go"
    "crypto/sha256"
)

var base64Secret = os.Getenv("APPROOV_BASE64_SECRET")

type SuccessResponse struct {
    Message string `json:"message"`
}

type ErrorResponse struct {}

func errorResponse(response http.ResponseWriter, statusCode int, message string) {
    response.Header().Set("Content-Type", "application/json")
    response.WriteHeader(statusCode)
    json.NewEncoder(response).Encode(ErrorResponse{})
}

func helloHandler(response http.ResponseWriter, request *http.Request) {
    response.Header().Set("Content-Type", "application/json")
    response.WriteHeader(http.StatusOK)
    json.NewEncoder(response).Encode(SuccessResponse{Message: "Hello, World!"})
}

func verifyApproovToken(response http.ResponseWriter, request *http.Request)  (*jwt.Token, error) {
    approovToken := request.Header["Approov-Token"]

    if approovToken == nil {
        return nil, fmt.Errorf("Token is missing.")
    }

    token, err := jwt.Parse(approovToken[0], func(token *jwt.Token) (interface{}, error) {

        if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
            return nil, fmt.Errorf("Signing method mismatch.")
        }

        secret, err := base64.StdEncoding.DecodeString(base64Secret)

        if err != nil {
            return nil, fmt.Errorf(err.Error())
        }

        return secret, nil
    })

    return token, err
}

func verifyApproovTokenBinding(token *jwt.Token, request *http.Request) (jwt.Claims, error) {
    claims := token.Claims
    token_binding_payload, has_pay_key := claims.(jwt.MapClaims)["pay"]

    if ! has_pay_key {
        // We don't return an error, because the Approov fail-over server doesn't provide the `pay` key.
        return claims, nil
    }

    // We use the Authorization token here, but feel free to use another header. However, you need to bind this header to the
    // Approov token in the mobile app.
    token_binding_header := request.Header["Authorization"]

    if len(token_binding_header) != 1 {
        return claims, fmt.Errorf("The header to perform the verification for the token binding is missing, empty or has more than one entry.")
    }

    // We need to hash and base64 encode the token binding header, because we need to compare it in the same way it was
    // included in the Approov token on the mobile app.
    token_binding_header_hashed := sha256.Sum256([]byte(token_binding_header[0]))
    token_binding_header_encoded := base64.StdEncoding.EncodeToString(token_binding_header_hashed[:])

    if token_binding_payload != token_binding_header_encoded {
        return claims, fmt.Errorf("Invalid token binding.")
    }

    return claims, nil
}

func makeApproovCheckerHandler(handler func(http.ResponseWriter, *http.Request)) http.Handler {
    return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {

        token, err := verifyApproovToken(response, request)

        if err != nil {
            errorResponse(response, http.StatusUnauthorized, err.Error())
            return
        }

        if ! token.Valid {
            errorResponse(response, http.StatusUnauthorized, "Token is invalid.")
            return
        }

        claims, err := verifyApproovTokenBinding(token, request)

        if err != nil {
            errorResponse(response, http.StatusUnauthorized, err.Error())
            return
        }

        // Use the returned claims to perform any additional checks.
        _ = claims

        handler(response, request)
    })
}

func main() {
    http.Handle("/", makeApproovCheckerHandler(helloHandler))
    log.Println("Server listening on http://localhost:8002")
    err := http.ListenAndServe(":8002", nil)
    if err != nil {
        log.Fatal("Server Error: " + err.Error())
    }
}

File: nodejs/src/backend-integration-impact/hello-server-token-binding-protected.js
// @link https://approov.io/docs/v2.0/approov-usage-documentation/#backend-integration-impact
// @link https://github.com/approov/documentation_backend-code-snippets/blob/master/nodejs/src/backend-integration-impact/hello-server-token-binding-protected.js

const express = require('express')
const jwt = require('express-jwt')
const crypto = require('crypto')
const app = express()

const dotenv = require('dotenv').config()

if (dotenv.error) {
  console.debug('FAILED TO PARSE `.env` FILE | ' + dotenv.error)
}

const SECRET = dotenv.parsed.APPROOV_BASE64_SECRET

const isEmpty = function(value) {
  return  (value === undefined) || (value === null) || (value === '')
}

const isString = function(value) {
  return (typeof(value) === 'string')
}

const isEmptyString = function(value) {
  return (isEmpty(value) === true) || (isString(value) === false) ||  (value.trim() === '')
}

const ERROR_RESPONSE_BODY = {}

// Callback that performs the Approov token check using the express-jwt library
const checkApproovToken = jwt({
  secret: Buffer.from(SECRET, 'base64'), // decodes the Approov secret
  requestProperty: 'approovTokenDecoded',
  getToken: function fromApproovTokenHeader(req) {
    req.approovTokenError = false
    return req.get('Approov-Token')
  },
  algorithms: ['HS256']
})

// Callback to handle the errors occurred while checking the Approov token.
const handlesApproovTokenError = function(err, req, res, next) {

  if (err.name === 'UnauthorizedError') {
    req.approovTokenError = true

    console.debug('APPROOV TOKEN ERROR: %s', err)

    res.status(401)
    res.json(ERROR_RESPONSE_BODY)
    return
  }

  next()
  return
}

// Callback to handles when an Approov token is successfully validated.
const handlesApproovTokenSuccess = function(req, res, next) {
  if (req.approovTokenError === false) {
    console.debug('VALID APPROOV TOKEN')
  }

  next()
  return
}

/**
 * @link https://approov.io/docs/v2.0/approov-usage-documentation/#backend-integration-impact
 * @link https://github.com/approov/documentation_backend-code-snippets/blob/master/nodejs/src/backend-integration-impact/hello-server-token-binding-protected.js
 */
const handlesApproovTokenBindingVerification = function(req, res, next) {

    // The decoded Approov token was added to the request object when the checked it at `checkApproovToken()`
    token_binding_payload = req.approovTokenDecoded.pay

    if (token_binding_payload === undefined) {
        console.debug("APPROOV TOKEN BINDING WARNING: key 'pay' is missing.")

        // We let the request to continue as usual, because the Approov fail-over server doesn't provide the `pay` key.
        next()
        return
    }

    if (isEmptyString(token_binding_payload)) {
        console.debug("APPROOV TOKEN BINDING ERROR: key 'pay' in the decoded token is missing or empty.")
        res.status(401)
        res.json(ERROR_RESPONSE_BODY)
        return
    }

    // We use here the Authorization token, but feel free to use another header, but you need to bind this  header to
    // the Approov token in the mobile app.
    token_binding_header = req.get('Authorization')

    if (isEmptyString(token_binding_header)) {
        console.debug("APPROOV TOKEN BINDING ERROR: Missing or empty header to perform the verification for the token binding.")
        res.status(401)
        res.json(ERROR_RESPONSE_BODY)
        return
    }

    // We need to hash and base64 encode the token binding header, because thats how it was included in the Approov
    // token on the mobile app.
    const token_binding_header_encoded = crypto.createHash('sha256').update(token_binding_header, 'utf-8').digest('base64')

    if (token_binding_payload !== token_binding_header_encoded) {
        console.debug("APPROOV TOKEN BINDING ERROR: Invalid token binding.")
        res.status(401)
        res.json(ERROR_RESPONSE_BODY)
        return
    }

    console.debug("VALID APPROOV TOKEN BINDING.")

    // Let the request continue as usual.
    next()
    return
}

// Intercepts all calls to the 'hello world' endpoint and validates the Approov token.
app.use('/', checkApproovToken)

// Handles failure in validating the Approov token
app.use('/', handlesApproovTokenError)

// Handles requests where the Approov token is a valid one.
app.use('/', handlesApproovTokenSuccess)

// Handles requests where the Approov token contains a bind to an another header in the request
app.use('/', handlesApproovTokenBindingVerification)

// simple 'hello world' endpoint.
app.get('/', function (req, res, next) {
    res.json({message: "Hello Approov token binding protected server!"})
})

// Create and run the HTTP server
app.listen(8002, function () {
  console.debug("Server listening on %s", "http://localhost:8002")
})
